### PR TITLE
Add support of user self signed certificates

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
A user can provides self signed certificates by putting them under .minio/certs/CAs/. I added a global variable which holds Root CAs. Unfortunately, under Windows, we can't load system + user supplied CAs at the same time. So a Windows user should supply all needed certificates.
 
## Motivation and Context
No issue was assigned

## How Has This Been Tested?
Manually

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [NOT YET] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/certs.go [hunk 1]
```diff
@@ -62,6 +66,26 @@ func mustGetKeyFile() string {
...
+	}
+	return
+}
+
+// mustGetSystemCertPool returns empty cert pool in case of error (windows)
+func mustGetSystemCertPool() *x509.CertPool {
+	pool, err := x509.SystemCertPool()
+	if err != nil {
+		return x509.NewCertPool()
+	} else {
```
Comment (harshavardhana) [inline cmd/certs.go]: this 'else block' is not needed.

File: cmd/server-main.go [hunk 2]
```diff
@@ -184,12 +185,34 @@ func finalizeEndpoints(tls bool, apiServer *http.Server) (endPoints []string) {
 	return endPoints
 }
 
+// loadRootCAs fetches CA files provided in minio config and add them to globalRootCAs
```
Comment (fwessels) [inline cmd/server-main.go]: ... and add**s** them ...

File: cmd/certs.go [hunk 3]
```diff
@@ -62,6 +66,25 @@ func mustGetKeyFile() string {
...
 }
 
+// mustGetCAFiles must get the list of the CA certificates stored in minio config dir
+func mustGetCAFiles() (caCerts []string) {
+	CAsDir := filepath.Join(mustGetCertsPath(), globalMinioCertsCADir)
+	caFiles, _ := ioutil.ReadDir(CAsDir)
+	for _, cert := range caFiles {
+		caCerts = append(caCerts, filepath.Join(CAsDir, cert.Name()))
+	}
+	return
```
Comment (fwessels) [inline cmd/certs.go:76]: `return` is a void statement here

**Comment by fwessels** [comment]
Apart from minor comments, LGTM

**Comment by harshavardhana** [comment]
This requires a rebase @vadmeste

**Comment by harshavardhana** [comment]
@vadmeste is this ready or still WIP?

**Comment by vadmeste** [comment]
> @vadmeste is this ready or still WIP?

Ready

**Comment by harshavardhana** [comment]
why not add a test case for generating self signed cert,  loading and making an RPC connection?

**Comment by vadmeste** [comment]
> why not add a test case for generating self signed cert, loading and making an RPC connection?

yeah, looks useful.. will do it tomorrow.

**Comment by harshavardhana** [comment]
> yeah, looks useful.. will do it tomorrow.

ping @vadmeste

**Comment by vadmeste** [comment]
> ping @vadmeste

starting on it right now.

File: docs/configure-minio-with-tls.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+```sh
+openssl genrsa -out root-ca.key 1024
+```
+
+Generate the self-signed certificate:
+```sh
+openssl req -new -x509 -days 3650 -key root-ca.key -out root-ca.crt -subj "/C=country/ST=state/L=location/O=organization/CN=domain"
+```
+
+### 3. Configure Minio with the generated certificate
```
Comment (harshavardhana) [inline docs/configure-minio-with-tls.md:39]: This heading should match the heading before should be `##`

File: docs/configure-minio-with-tls.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+```sh
+openssl req -new -x509 -days 3650 -key root-ca.key -out root-ca.crt -subj "/C=country/ST=state/L=location/O=organization/CN=domain"
+```
+
+### 3. Configure Minio with the generated certificate 
+
+To make Minio aware about your generated key and certificate, you will need to put them under `certs` directory in your Minio config path (usually ~/.minio) using the names of `private.key` and `public.crt` for key and certificate files respectively.
+
+
+### 4. Install third parties CAs
```
Comment (harshavardhana) [inline docs/configure-minio-with-tls.md:44]: Same as above.

File: docs/configure-minio-with-tls.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+```sh
+  go run generate_cert.go -ca --host "10.10.0.3"
+```
+
+### With OpenSSL:
+
+Generate the private key:
+```sh
+openssl genrsa -out root-ca.key 1024
```
Comment (harshavardhana) [inline docs/configure-minio-with-tls.md:31]: Keep the names as `private.key`

File: docs/configure-minio-with-tls.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+```
+
+Generate the self-signed certificate:
+```sh
+openssl req -new -x509 -days 3650 -key root-ca.key -out root-ca.crt -subj "/C=country/ST=state/L=location/O=organization/CN=domain"
```
Comment (harshavardhana) [inline docs/configure-minio-with-tls.md:36]: Keep the names as `private.key` and `public.crt`

File: docs/configure-minio-with-tls.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+Minio supports only key/certificate in PEM format. 
+
+### With Letsencrypt
+
+https://docs.minio.io/docs/generate-let-s-encypt-certificate-using-concert-for-minio
+
+### With generate_cert.go (self-signed certificate)
+
+You need to download https://golang.org/src/crypto/tls/generate_cert.go?m=text which is a simple go tool for generating self-signed certificates but works for the most of cases.
```
Comment (harshavardhana) [inline docs/configure-minio-with-tls.md:19]: Make this linked entity by using something like 

```
[download](https://golang.org/src/crypto/tls/generate_cert.go?m=text)
```

File: docs/configure-minio-with-tls.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+* Download Minio server from [here](https://docs.minio.io/docs/minio)
+
+## 2. Generate TLS certificate
+
+Minio supports only key/certificate in PEM format. 
+
+### With Letsencrypt
+
+https://docs.minio.io/docs/generate-let-s-encypt-certificate-using-concert-for-minio
```
Comment (harshavardhana) [inline docs/configure-minio-with-tls.md:15]:
```
Please explore [here](https://docs.minio.io/docs/generate-let-s-encypt-certificate-using-concert-for-minio)
```

**Comment by harshavardhana** [comment]
Merged in master - closing.

**Pull request closed without merging**